### ISSUE TITLE
tensorflow-gpu error in Chapter 5

### ISSUE THREAD

**Issue by Grippr** [PacktPublishing/Neural-Network-Projects-with-Python#9]
I'm having a problem in chapter 5 where I get the following error: 

<snip>
Epoch 1/10
2020-11-02 13:40:42.502689: I tensorflow/core/platform/cpu_feature_guard.cc:142] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX AVX2
2020-11-02 13:40:42.507971: I tensorflow/stream_executor/platform/default/dso_loader.cc:42] Successfully opened dynamic library nvcuda.dll
2020-11-02 13:40:42.629293: E tensorflow/stream_executor/cuda/cuda_driver.cc:318] failed call to cuInit: CUDA_ERROR_NO_DEVICE: no CUDA-capable device is detected
2020-11-02 13:40:42.631779: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:169] retrieving CUDA diagnostic information for host: DESKTOP-LU2D9D5
2020-11-02 13:40:42.631939: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:176] hostname: MY_DESKTOP
</snip> 

For reference, when I run the chapter04 auto encoder, it works fine with the gpu. 
I also tried adding batch_size=32 to the fit, the code runs, but on the cpu. 

<snip>
Epoch 1/10
2020-11-02 13:42:15.445419: I tensorflow/core/platform/cpu_feature_guard.cc:142] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX AVX2
2020-11-02 13:42:15.449756: I tensorflow/stream_executor/platform/default/dso_loader.cc:42] Successfully opened dynamic library nvcuda.dll
2020-11-02 13:42:15.563797: E tensorflow/stream_executor/cuda/cuda_driver.cc:318] failed call to cuInit: CUDA_ERROR_NO_DEVICE: no CUDA-capable device is detected
2020-11-02 13:42:15.566350: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:169] retrieving CUDA diagnostic information for host: DESKTOP-LU2D9D5
2020-11-02 13:42:15.566498: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:176] hostname: DESKTOP-LU2D9D5
195/195 [==============================] - 11s 58ms/step - loss: 0.4456
</snip?

**Comment by Grippr** [comment]
I...got this to work eventually, here is the updated code:

import matplotlib
matplotlib.use("TkAgg")
from keras.datasets import mnist
from keras.models import Sequential
from keras.layers import Dense
from matplotlib import pyplot as plt
import numpy as np
import random

import tensorflow as tf
config = tf.ConfigProto( device_count = {'GPU': 1 , 'CPU': 20} )
config.gpu_options.allow_growth = True
config.gpu_options.per_process_gpu_memory_fraction = 1.0
sess = tf.Session(config=config) 
from keras import backend
backend.set_session(sess)

def create_basic_autoencoder(hidden_layer_size):
  model = Sequential() 
  model.add(Dense(units=hidden_layer_size, input_shape=(784,), activation='relu'))
  model.add(Dense(units=784, activation='sigmoid'))
  return model
#
if __name__ == '__main__':
      # Import MNIST dataset
  training_set, testing_set = mnist.load_data()
  X_train, y_train = training_set
  X_test, y_test = testing_set

  # Reshape the dataset for our neural network
  X_train_reshaped = X_train.reshape((X_train.shape[0], X_train.shape[1]*X_train.shape[2]))
  X_test_reshaped = X_test.reshape((X_test.shape[0], X_test.shape[1]*X_test.shape[2]))

  # Normalize range of values between 0 to 1 (from 0 to 255)
  X_train_reshaped = X_train_reshaped/255.
  X_test_reshaped = X_test_reshaped/255.

  # Add noise to the MNIST dataset
  X_train_noised = X_train_reshaped + np.random.normal(0, 0.5, size=X_train_reshaped.shape) 
  X_test_noised = X_test_reshaped + np.random.normal(0, 0.5, size=X_test_reshaped.shape)
  X_train_noised = np.clip(X_train_noised, a_min=0, a_max=1)
  X_test_noised = np.clip(X_test_noised, a_min=0, a_max=1)

  # Model Building and Training
  basic_denoise_autoencoder = create_basic_autoencoder(hidden_layer_size=16)
  basic_denoise_autoencoder.compile(optimizer='adam', loss='mean_squared_error')
  basic_denoise_autoencoder.fit(X_train_noised, X_train_reshaped, epochs=10)
  output = basic_denoise_autoencoder.predict(X_test_noised)

    # Plot output
  fig, ((ax1, ax2, ax3, ax4, ax5), (ax6, ax7, ax8, ax9, ax10), (ax11,ax12,ax13,ax14,ax15)) = plt.subplots(3, 5)
  randomly_selected_imgs = random.sample(range(output.shape[0]),5)

  # 1st row for original images
  for i, ax in enumerate([ax1,ax2,ax3,ax4,ax5]):
    ax.imshow(X_test_reshaped[randomly_selected_imgs[i]].reshape(28,28), cmap='gray')
    if i == 0:
      ax.set_ylabel("Original \n Images")
    ax.grid(False)
    ax.set_xticks([])
    ax.set_yticks([])

  # 2nd row for input with noise added
  for i, ax in enumerate([ax6,ax7,ax8,ax9,ax10]):
    ax.imshow(X_test_noised[randomly_selected_imgs[i]].reshape(28,28), cmap='gray')
    if i == 0:
      ax.set_ylabel("Input With \n Noise Added")
    ax.grid(False)
    ax.set_xticks([])
    ax.set_yticks([])

  # 3rd row for output images from our autoencoder
  for i, ax in enumerate([ax11,ax12,ax13,ax14,ax15]):
    ax.imshow(output[randomly_selected_imgs[i]].reshape(28,28), cmap='gray')
    if i == 0:
      ax.set_ylabel("Denoised \n Output")
    ax.grid(False)
    ax.set_xticks([])
    ax.set_yticks([])

  plt.show()

**Comment by Grippr** [comment]
I was also having problems with the noisy documents, but I was able to get rid of the problem by setting batch_size=8 in the fit
